### ISSUE TITLE
About using compatibility patches and ModCheck

### ISSUE THREAD

**Issue by kaptain-kavern** [spoonshortage/A_Dog_Said#3]
In order to not pollute forum threads more, I made this issue for discussing code. _With hope/luck, we even may end up with some more documentation or pre-made patch mod like I did in the other issue_

**Comment by Nightinggale** [comment]
Regarding the discussion about the way to patch modded animals into supporting A Dog Said...

There are apparently two approaches:
1: patch each animal with a list of recipes
2: patch the recipes with a list of animals

Both should apparently work just fine, but the main question is: which one is the fastest?
It's the one, which is best at avoiding slow code. Like most other patches, the amount of actual modification to each xml file is minimal and RimWorld is using xml from System, which we can assume to be highly optimized. In other words the actual patching is instant and we can forget about it. xPath on the other hand is not instant.

This brings up the next question: how many xPath searches will a patch file make? Each patch file is actually applied to every single def xml file in every single mod. Core alone brings the def filecount to 391 and it doesn't take much effort to add mods to exceed 1000. The more mods the game uses, the more important it is to have each one working as fast as possible.

Conclusion:
# **The fastest patch file is the one with the fewest xPath searches in files where it won't do anything**

The [ADS patching guide](https://github.com/spoonshortage/A_Dog_Said/issues/1) will always run 21 xPath searches. With 21k searches, it doesn't really matter how much is done the 21 times each is actually used.

Patching each animal seem to be done as one xPath search for each animal, meaning the two approaches are equally fast if there are 21 animals.

However we can do much better than that!

Let's skip OldWoundsAnimal for a moment.

Nextrecipe to patch is InstallBionicAnimalStomach. After that it will search for InstallBionicAnimalKidney. However if InstallBionicAnimalStomach fail, we know it's not Recipes_Surgery.xml in ADS, meaning InstallBionicAnimalKidney will not be there. It would be faster to not search for it.

PatchOperationSequence will stop if a PatchOperation in the list fails. We should make a sequence, which first patches InstallBionicAnimalStomach, then InstallBionicAnimalKidney. If none of those are in a file, we saved an xPath search. Follow this system and we can bring the number of searches in "wrong" files down to the number of files we need to patch, which for ADS appears to be 3. That's 7 times faster than the guide.

We can do even more than that. If we really aim for the fastest possible patching, we can modify ADS to join the 3 recipe files into one, in which case all of them can be handled by one sequence and then only one xPath search is needed in other files, which would **make patching 21 times faster. That's a 95.2% reduction of time spend for each mod patching for ADS.**
It would still need 21 searches in the file in question, but if it searches 1200 or 1221 doesn't really make a difference. 1221 or 3600 or 25200 will.

A sequence will be needed anyway now that mods are starting to use ModCheck.isModLoaded. Like all other calls to ModCheck, this one will execute the first time, then save the result in memory and use that for the following calls. As a result, from a time perspective it is only executed once regardless of the number of def files. This mean patching for ADS if ADS isn't loaded will not slow down the game startup.

The next question is what to write in xPath to make each search as fast as possible. The answer to this is pending because while I have an idea, I have just come up with a way to make a test where it can actually be measured.


None of this is about my idea where PatchOperationAdd can be looped with multiple xPaths. However after thinking about it, writing all this and doing a bit of research to get it right, I think such a group approach is wrong. While it will make it easier and faster to make patch files, it will encourage people to use more xPath searches. If anything is added to ModCheck, it should either not affect the number of xPath searches or reduce it.

**Comment by Nightinggale** [comment]
I decided to write a summary of the wall of text. The different approaches to patching to add ADS support measured in time units.

- Patch each animal: 1 unit for each animal
- Using template: 21 units
- Optimizing template: 3 units
- Moving all ADS recipes into one file and then patch with optimized template: 1 unit

Particularly due to the number of mods patching for ADS, it would be preferable to make it patch as fast as possible. It really comes down to nice xml files vs performance. If the recipe def files are merged into one, then patching can be 3 times as fast, but the recipe file will be big. Personally I would go with the high performance option, but it's up to you.

I plan to write an optimized template, which also makes use of ModCheck*, but I will need to know if it should work on 1 or 3 recipe files before starting.
*ModCheck means the patch can be added to the mod with the animals and then it will patch the animals if ADS is loaded. If not, then the patch will be ignored.

**Comment by spoonshortage** [comment]
Please explain to me what kind of times we are actually looking at. (I was under the impression people would just use patchoperationfindmod as soon as that one is fixed and patched)
so what actual real life time are we looking at? is it something like .1 seconds vs 30 seconds or times like .1 vs .8 (which is not even really noticeable imho xD, also i pulled those times out of my ass here i have really no idea) and what exactly would I have to change in my files to accomodate animal mod makers better? I would not like to join my recipe xml files tbh bc that is just horribly unpractical for myself.

**Comment by Nightinggale** [comment]
I did some systematic measurements to get more precise numbers to talk about. Each "time unit" is hard to translate into seconds because not only will it highly depend on the hardware, it will also depend on the number of def xml files and the size of each, meaning it depends on which mods are loaded. Because of this, it makes more sense to talk about a baseline without patches and then how much a time unit adds in % of the base.

My measurements reveal that we are talking about 0.7-1% increase in xml loading time for each time unit. Note that it stacks by addition, not multiplication, meaning 10 time units will add 7-10%. This mean 5 mods using the current template will double the loading time, but we already knew we would want to do better than that.

This mean it's a question of if it should be 0.7-1% or 2-3% increase in loading time for each mod patching for ADS, or an additional 1.5% for not merging the recipes. That's a little under a second for every minute of the base time.

I will say it's enough to consider because it stacks with each mod using ADS, but at the same time it's likely not enough to justify making it noteworthy harder to work on ADS itself. 

I will say we should stick with 3 files. It's a tradeoff, we examined the pro and con for each option and made the decision based on that.


And about patchoperationfindmod 
I'm not sure what will happen to it. It's flamed for not really helping with how mods usually work. However it matches perfectly with vanilla internals. For the time being, I say we should assume people will use ModCheck and if patchoperationfindmod will work well in the future, then it can replace ModCheck in the template. It should be noted that ModCheck has an internal cache to make it work much faster while no vanilla patch operation has that. In fact I think ModCheck is the only PatchOperation to use caching.
